test/test_curl.cpp

```diff
@@ -344,5 +373,65 @@ TEST_CASE("setopt failures") {
   REQUIRE(result.error().code == Error::CURL_REQUEST_SETUP_FAILED);
 }
 
+TEST_CASE("handles are always cleaned up") {
+  const auto logger = std::make_shared<MockLogger>();
+  SingleRequestMockCurlLibrary library;
+  auto client = std::make_shared<Curl>(logger, library);
+
+  SECTION("when the response is delivered") {
+    Optional<Error> post_error;
+    std::exception_ptr exception;
+    const HTTPClient::URL url = {"http", "whatever", ""};
+    const auto result = client->post(
+        url, [](const auto &) {}, "whatever",
+        [&](int status, const DictReader & /*headers*/, std::string body) {
+          try {
+            REQUIRE(status == 200);
+            REQUIRE(body ==
+                    "{\"message\": \"Dogs don't know it's not libcurl!\"}");
+          } catch (...) {
+            exception = std::current_exception();
+          }
+        },
+        [&](const Error &error) { post_error = error; });
+
+    REQUIRE(result);
+    client->drain(std::chrono::steady_clock::now() + std::chrono::seconds(1));
+    if (exception) {
+      std::rethrow_exception(exception);
+    }
+    REQUIRE_FALSE(post_error);
+  }
+
+  SECTION("when an error occurs") {
+    Optional<Error> post_error;
+    const HTTPClient::URL url = {"http", "whatever", ""};
+    const auto ignore = [](auto &&...) {};
+    library.message_result_ = CURLE_COULDNT_CONNECT;  // any error would do
+    const auto result =
+        client->post(url, ignore, "whatever", ignore,
+                     [&](const Error &error) { post_error = error; });
+
+    REQUIRE(result);
+    client->drain(std::chrono::steady_clock::now() + std::chrono::seconds(1));
+    REQUIRE(post_error);
+  }
+
+  SECTION("when we shut down while a request is in flight") {
+    const HTTPClient::URL url = {"http", "whatever", ""};
+    const auto ignore = [](auto &&...) {};
+    library.delay_message_ = true;
+    const auto result = client->post(url, ignore, "whatever", ignore, ignore);
+
+    REQUIRE(result);
+    // Destroy the `Curl` object.
+    client.reset();
+  }
+
+  // Here are the checks relevant to this test.
+  REQUIRE(library.created_handles_.size() == 1);
+  REQUIRE(library.created_handles_ == library.destroyed_handles_);
```

Comment:
The `SECTION`s above are all setup. The post-condition, here, is always the same: one handle created and one handle destroyed.